common/searchattribute/validator.go

```diff
@@ -96,14 +96,15 @@ func (v *Validator) Validate(searchAttributes *commonpb.SearchAttributes, namesp
 	}
 
 	for saName, saPayload := range searchAttributes.GetIndexedFields() {
-		if IsReservedField(saName) {
-			return serviceerror.NewInvalidArgument(fmt.Sprintf("%s is Temporal reserved field name", saName))
+		if !typeMap.IsDefined(saName) {
+			return serviceerror.NewInvalidArgument(fmt.Sprintf("%s is not a valid search attribute name", saName))
 		}
 
-		saType, err := typeMap.GetType(saName)
+		saType, err := typeMap.getType(saName, customCategory|predefinedCategory)
 		if err != nil {
-			return serviceerror.NewInvalidArgument(fmt.Sprintf("%s is not a valid search attribute name", saName))
+			return serviceerror.NewInvalidArgument(fmt.Sprintf("%s attribute can't be set in SearchAttributes: reserved field name", saName))
```

Comment:
is reserved filed name the only possible reason this could fail?

Reply:
yes, but I agree this code was confusing. I reorganized it for clarity.